Deprecate override of stringPrefix

ashawley opened the pull request:
While working on updating scala-xml to 2.13.0-M4, I ran across a problem with its overriding `stringPrefix`:

```
[error]   scala-xml/shared/src/main/scala/scala/xml/NodeBuffer.scala:24:
[error]   overriding method stringPrefix in trait IterableOps of type => String;
[error]   method stringPrefix cannot override final member
[error]   override def stringPrefix: String = "NodeBuffer"
```

Looks like it was marked `deprecated` in 2.13.0, but it's been changed to `final`, so any existing code that overrides it will no longer work.

Is the plan to remove it in 2.14.0?  My guess is nobody is actually calling `stringPrefix`, since the point is to override it with something else so that it is used by `toString`.

julienrf reviewed (changes requested):
The new name for this method is `className`. Overriding `stringPrefix` has no effect. If you remove the `final` qualifier on `stringPrefix` then people may accidentally override `stringPrefix` without noticing that this override is wrong.

What we could do:

1. Rename back `className` to `stringPrefix` to get rid of the incompatibility,
2. Or, remove `final` and replace the `@deprecated` with `@deprecatedOverriding` to turn the error you got into only a warning.

som-snytt commented:
It seems more migration-friendly to both delegate `className` to `stringPrefix` and use `@deprecatedOverriding` instead of `final`.

sjrd commented:
TBH, I have lost hope in the ability to cross-compile own collections. You just need to use separate source directories.

`className` vs `stringPrefix` is only one out of many methods that need to be overridden differently. For example, `+=` versus `addOne`, `-=` versus `subtractOne`, and many others.

SethTisue commented:
> TBH, I have lost hope in the ability to cross-compile own collections

yeah, we weren't even trying to support that.

ashawley commented:
> If you remove the `final` qualifier on `stringPrefix` then people may accidentally override `stringPrefix` without noticing that this override is wrong.

Yeah, only later did I realize this change would be more involved than just removing `final`.

> It seems more migration-friendly to both delegate `className` to `stringPrefix` and use `@deprecatedOverriding` instead of `final`.

This is a useful suggestion, I gave it a try.

I suppose it raises a few issues, though.  

1. It's not clear what the behavior is when you override both fields, but I suppose no one would be able to actually do this to for the purpose of making their code cross-compile.
1. When `stringPrefix` is removed in 2.14, it seems like it will just be a hard break anyway?

> The new name for this method is `className`.

So what's the point of keeping this field, `stringPrefix`, around then?  The error, "`cannot override final member`" is not very helpful, IMO, since I presume most people `override` the field.

Should we just delete this field and retroactively have it marked deprecated in 2.12?

SethTisue commented:
> Should we just delete this field

👍

> and retroactively have it marked deprecated in 2.12?

I don't think we need to do that, we don't deprecate things unless a replacement exists (or unless the thing simply isn't recommended for use at all), but in 2.12 no replacement exists for people writing their own collections.

sjrd commented:
+1 to what @SethTisue said. In fact we *must not* deprecate `stringPrefix` in 2.12, since it is clearly used and has no replacement.

I would like to consider making `className` `protected[this]`. There should be no reason to call it from outside the class itself. (yes, it's a breaking change)

ashawley commented:
> we don't deprecate things unless a replacement exists

Is this the case?  Isn't `className` the replacement?

> I would also like to consider making `className` `protected[this]`.

Yeah, +1 to that.

SethTisue commented:
> > we don't deprecate things unless a replacement exists

> Is this the case? Isn't className the replacement?

unless a replacement exists in the same Scala version in which we are deprecating the old thing.

ashawley commented:
Ah, ok.  Too bad we can't retroactively add `className` to 2.12.

sjrd commented:
Also worth considering: this change is not compatible for *clients* of the `stringPrefix` method. Indeed, they won't see the behavior in overridden `className`s anymore. Therfore this change slightly improves custom collections bit worsens clients of the new collections.

Given that compatibility for custom collections is not really a goal, but that of clients is, it seems this PR is going against the started goal.

julienrf commented:
Yes, good point @sjrd, I forgot that part about clients of `stringPrefix`. Then we are back to the solution proposed [above](https://github.com/scala/scala/pull/6659#pullrequestreview-120937303)?

> remove `final` and replace the `@deprecated` with `@deprecatedOverriding` to turn the error you got into only a warning.

ashawley commented:
> this change is not compatible for _clients_ of the `stringPrefix` method.

Are there any _clients_ of `stringPrefix` (external to the compiler itself)?

ashawley commented:
> There should be no reason to call it from outside the class itself.

It turns out external calls to `className` are made by `ScalaRunTime`, so I was only able to mark it as `protected[scala]` and not `protected[this]`.

sjrd commented:
The requirements of `ScalaRunTime`, whatever they are, should not leak in the public or protected API. Consider making the method `protected[this]` anyway, and provide a `private[scala]` *forwarder* for the exclusive use of `ScalaRunTime` (comment it as such).

ashawley commented:
Ok, I tried to add a _forwarder_ as you suggested, and called it `collectionClassName`.  It is used by various tests in the test suite, as well.

lrytz commented on src/library/scala/collection/Iterable.scala:
```diff
@@ -194,7 +194,19 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
+    * `scala.runtime.ScalaRunTime.stringOf` (and a few tests in
+    * the test suite).
+    */
+  private[scala] def collectionClassName: String = className
```
This should probably be `final`

ashawley replied:
Yes, it probably should and could be.  I've made it so.

lrytz commented:
LGTM, thanks! Could you squash all the changes into a single commit and summarize the change in the commit message, and also the PR description?

ashawley commented:
Can do.  How does this sound for a summary?:

- "Deprecate override of stringPrefix"

ashawley commented:
I squashed and rebased.  The tests are pending.

All that is left is changing the PR title.

ashawley committed:
Deprecate override of stringPrefix

The public method, stringPrefix, for collections was initially
released in 2.13.0-M4 as deprecated and marked final.  Since
most users override this field in 2.12 and earlier, it would
produce the compiler error:

  cannot override final member

but not the deprecation warning,

  method stringPrefix is deprecated in trait Iterable (since
  2.13.0): Use className instead of stringPrefix

It would be more migration-friendly to delegate the new
method, className, to the old method, stringPrefix.  The
annotation should also have been deprecatedOverriding and not
deprecated.  That way users will see the deprecation warning:

  overriding method stringPrefix in trait Iterable is
  deprecated (since 2.13.0): Override className instead

The final keyword on stringPrefix is dropped, but the new
Iterable.className and stringPrefix are marked as protected.

Clients of a collection will no longer have access to the old
or new field. A new final field, collectionClassName, is added
as a private[scala] forwarder so that className has the proper
visibility, but the forwarder provides the exclusive access
needed by scala.runtime.ScalaRunTime#stringOf (and also a few
tests in the test suite).

lrytz reviewed (approved):
Thank you!
